8292266: Add JFR event for total Reference processing time

sangheon opened the pull request:
Hi all,

Can I have reviews for this change that adds a new JFR event for total Reference Processing time?

The goal is to generate a new JFR event if there is already measured time. This CR doesn't add new time measurement.

Testing: hs-tier1 ~ 3

Thanks,
Sangheon

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292266](https://bugs.openjdk.org/browse/JDK-8292266): Add JFR event for total Reference processing time


### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [ec9209a5](https://git.openjdk.org/jdk/pull/11230/files/ec9209a571e5efeaf1cbd80cbb8879c555161ad8)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [2b119e74](https://git.openjdk.org/jdk/pull/11230/files/2b119e7490af78f6fea3384423f1399a14168893)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11230/head:pull/11230` \
`$ git checkout pull/11230`

Update a local copy of the PR: \
`$ git checkout pull/11230` \
`$ git pull https://git.openjdk.org/jdk pull/11230/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11230`

View PR using the GUI difftool: \
`$ git pr show -t 11230`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11230.diff">https://git.openjdk.org/jdk/pull/11230.diff</a>

</details>

sangheon commented:
/label add hotspot-gc, hotspot-jfr
/label remove hotspot

tschatzl commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -390,6 +390,12 @@
     <Field type="ulong" name="count" label="Total Count" />
   </Event>
 
+  <Event name="GCReferenceProcessTime" category="Java Virtual Machine, GC, Reference" label="GC Reference Process Time" startTime="false"
```
```suggestion
  <Event name="GCReferenceProcessingTime" category="Java Virtual Machine, GC, Reference" label="GC Reference Processing Time" startTime="false"
```

tschatzl commented on src/hotspot/share/jfr/metadata/metadata.xml:
```diff
@@ -390,6 +390,12 @@
   </Event>
 
+  <Event name="GCReferenceProcessTime" category="Java Virtual Machine, GC, Reference" label="GC Reference Process Time" startTime="false"
+    description="Total processing time of processed references during GC">
```
```suggestion
    description="Total time for processing java.lang.ref.References during GC">
```

tschatzl reviewed (approved):
Some minor comments in explanation texts/naming, but lgtm.

albertnetymk commented on src/hotspot/share/gc/shared/gcTrace.cpp:
```diff
@@ -72,6 +72,8 @@ void GCTracer::report_gc_reference_stats(const ReferenceProcessorStats& rps) con
   send_reference_stats_event(REF_FINAL, rps.final_count());
   send_reference_stats_event(REF_PHANTOM, rps.phantom_count());
+
+  send_reference_process_time_event(rps.total_process_time());
```
I think this is called by all collectors, but Z and Shenandoah would have zero for "total process time" here. I am not sure if this causes confusion on reading the jfr reports.

sangheon replied:
You are right and this is current scope of this CR.
If both collectors measure spent time, we can add it later. Hopefully under separate CR.

Or if the confusion only comes from zero value of the measured time, we can avoid sending zero value event. i.e. skip sending ref-proc-time event for both collectors. 
My first commit is something like this but stopped as someone may complain lack of the event anyway.

egahlin replied:
> > Or if the confusion only comes from zero value of the measured time
> 
> Yes, that's the confusion I was referring to.
> 
> > My first commit is something like this but stopped as someone may complain lack of the event anyway.
> 
> I see; I actually like the approach in your first commit better -- separate methods for reporting ref counts and processing time. If an event is not supported for a particular collector, it should not be emitted for that collector, IMO.
> 
> I still feel the behavior, reporting zero for ref-processing time for Shenandoah and Z, can be surprising, so I'd prefer this is avoided. However, if others think this is fine, I would not block this.

I agree about the zeros.

I also prefer that events are implemented for all collectors. It confusing for users or tools that consume the data to not get it for all GCs. I understand it is more work, but I rather delay the event a release than having something that is only partly done for JDK 20.

albertnetymk reviewed (approved):
> Or if the confusion only comes from zero value of the measured time

Yes, that's the confusion I was referring to.

> My first commit is something like this but stopped as someone may complain lack of the event anyway.

I see; I actually like the approach in your first commit better -- separate methods for reporting ref counts and processing time. If an event is not supported for a particular collector, it should not be emitted for that collector, IMO.

I still feel the behavior, reporting zero for ref-processing time for Shenandoah and Z, can be surprising, so I'd prefer this is avoided. However, if others think this is fine, I would not block this.

sangheon commented:
1. Reverted to the first version which has separate GCTrace::report_gc_reference_processing_time().
2. Added Z and Shenandoah impl. 

For Z and Shenandoah, I understand your point. But this is why I explicitly mentioned on CR/PR that I intended to emit already measured TOTAL time via ReferenceProcessorStats class. i.e. not add something new.

Thanks @albertnetymk and @egahlin for your comments.

tschatzl commented:
I'm fine with either solution (i.e. to not send the event for not-supported collectors), but really, it depends on what this new event is supposed to capture.

I.e. the question is what the "total reference processing time" is supposed to mean: is it reference processing time in stw pauses? Or generally "total processing time" (including cpu time spent concurrently)?

To me it's the former, so a value of 0.0 is technically correct for the concurrent collectors (actually I'm not sure whether Shenandoah does fully concurrent reference processing). This would suggest to make it available for all collectors, as there is a difference between not having a value for some collectors as it may not only mean "the time is zero" (does not do stw reference processing) but also something like "the collector does not do this at all".

When/if we add concurrent reference processing time tracking (which isn't trivial to do), then it may be useful to add another event (or just a value to the existing one) to be able to discriminate between concurrent vs. time spent in stw.

tschatzl commented:
> When/if we add concurrent reference processing time tracking (which isn't trivial to do), then it may be useful to add another event (or just a value to the existing one) to be able to discriminate between concurrent vs. time spent in stw.

Maybe name the value in the event something like "STW time" or so to make it clear already what it contains.

albertnetymk commented:
I think the new event for ref-processing time should be roughly corresponding to its neighbor `<Event name="GCReferenceStatistics"` in `metadata.xml`, regardless of ref-processing being STW or not.

> When/if we add concurrent reference processing time tracking (which isn't trivial to do)

@sangheon 's latest revision does that for Shenandoah and Z.

tschatzl commented:
> I think the new event for ref-processing time should be roughly corresponding to its neighbor `<Event name="GCReferenceStatistics"` in `metadata.xml`, regardless of ref-processing being STW or not.

GCReferenceStatistics isn't dependent on whether the work has been performed in stw or concurrently; it shows number of the processed references in the last "pass".

> 
> > When/if we add concurrent reference processing time tracking (which isn't trivial to do)
> 
> @sangheon 's latest revision does that for Shenandoah and Z.

Okay, but addition of this is exactly what makes the event misleading (or hard to explain to users) as I tried to explain: the Shenandoah and Z timings are different in the circumstances they are taken:
* for ZGC and Shenandoah it means wall time for concurrent reference processing. This can be arbitrarily lengthened due to mutator impact.
* for Serial/Parallel it means wall time for stw reference processing. Which is probably fairly close to cpu time given it is done with the mutator suspended.
* for G1 which does both, concurrent weak/reference processing work (preclean) and stw reference processing, this event now only tracks the latter. Adding them together for this event obviously does not make sense. Maybe G1 is the odd one out?

I.e. the values between collectors are not comparable although they use the same event/value, and even missing out some parts now.

This explains my reluctance to mix different kind of work (concurrent vs. stw) in a single event/value. This can be really confusing (apart from losing G1's concurrent reference processing work here). That is the reason I liked the original limit of this PR's scope to just the stw part, even if that value was always zero for some collectors due to the way they work, better.

(All of these timings do not cover time spent in discovery, which is arguably also some kind of "reference processing", but probably not something this event wants to track)

tschatzl commented:
Having said that, if everyone else agrees with having only one event for whatever a particular GC calls "reference processing", then I'm fine with it.

fisk reviewed (approved):
Looks good to me.

albertnetymk commented:
> This explains my reluctance to mix different kind of work (concurrent vs. stw) in a single event/value.

I acknowledge all the drawbacks of having a single event for ref-processing, but I believe it's a good trade-off, a simpler interface that captures the most significant component in most cases.
